testing/unstable_stub_test.ts

```diff
@@ -0,0 +1,394 @@
+    MockError,
+    "Cannot restore: instance method already restored",
+  );
+  assertEquals(func.restored, true);
+});
+
+// This doesn't test any runtime code, only if the TypeScript types are correct.
+Deno.test("stub() correctly handles types", () => {
+  // @ts-expect-error Stubbing with incorrect argument types should cause a type error
+  stub(new Point(2, 3), "explicitTypes", (_x: string, _y: number) => true);
+
+  // @ts-expect-error Stubbing with an incorrect return type should cause a type error
+  stub(new Point(2, 3), "explicitTypes", () => "string");
+
+  // Stubbing without argument types infers them from the real function
+  stub(new Point(2, 3), "explicitTypes", (_x, _y) => {
+    // `toExponential()` only exists on `number`, so this will error if _x is not a number
+    _x.toExponential();
+    // `toLowerCase()` only exists on `string`, so this will error if _y is not a string
+    _y.toLowerCase();
+    return true;
+  });
+
+  // Stubbing with returnsNext() should not give any type errors
+  stub(new Point(2, 3), "explicitTypes", returnsNext([true, false, true]));
+
+  // Stubbing without argument types should not cause any type errors:
+  const point2 = new Point(2, 3);
+  const explicitTypesFunc = stub(point2, "explicitTypes", () => true);
+
+  // Check if the returned type is correct:
+  assertThrows(() => {
+    assertSpyCall(explicitTypesFunc, 0, {
+      // @ts-expect-error Test if passing incorrect argument types causes an error
+      args: ["not a number", "string"],
+      // @ts-expect-error Test if passing incorrect return type causes an error
+      returned: "not a boolean",
+    });
+  });
+
+  // Calling assertSpyCall with the correct types should not cause any type errors
+  point2.explicitTypes(1, "hello");
+  assertSpyCall(explicitTypesFunc, 0, {
+    args: [1, "hello"],
+    returned: true,
+  });
+});
+
+Deno.test("stub() works with throwing fake implementation", () => {
+  const obj = {
+    fn() {
+      throw new Error("failed");
+    },
+  };
+  const stubFn = stub(obj, "fn", () => {
+    throw new Error("failed");
+  });
+  assertThrows(() => obj.fn(), Error, "failed");
+  assertSpyCall(stubFn, 0, {
+    self: obj,
+    args: [],
+    error: { Class: Error, msgIncludes: "failed" },
+  });
+});
+
+Deno.test("stub() throws when the property is not a method", () => {
+  const obj = { fn: 1 };
+  assertThrows(
+    // deno-lint-ignore no-explicit-any
+    () => stub(obj as any, "fn"),
+    MockError,
+    "Cannot stub: property is not an instance method",
+  );
+});
+
+Deno.test("stub() throws when try stubbing already stubbed method", () => {
+  const obj = { fn() {} };
+  stub(obj, "fn");
+  assertThrows(
+    () => stub(obj, "fn"),
+    MockError,
+    "Cannot stub: already spying on instance method",
+  );
+});
+
+Deno.test("stub() throws when neither setter not getter is defined", () => {
+  const obj = { prop: "foo" };
+
+  assertThrows(
+    () => stub(obj, "prop", {}),
```

Comment:
I think it would make sense to support an empty property descriptor for when you want it to start with the current value. I think we should also support passing a value option instead of a getter/setter in cases where they want to both spy on and override the value. Without this suggested change, you'd have to create getters and setters like this, where x could be the current value or the value you want to override it with.

Without this suggestion:

```ts
const point = new Point(5, 6);
let x = point.x;
const prop = stub(point, "x", {
  get: function () {
    return x;
  },
  set: function (value: number) {
    x = value;
  },
});
```

With this suggestion when you want to spy without changing the initial value you would do this:

```ts
const point = new Point(5, 6);
const prop = stub(point, "x", {});
```

If possible to get the typing right, it would be even cooler if you could stub a property like you can stub a function, without the third arg.

```ts
const point = new Point(5, 6);
const prop = stub(point, "x");
```

With this suggestion when you want to spy while also changing the initial value:

```ts
const point = new Point(5, 6);
const prop = stub(point, "x", { value: 3 });
```

With this suggestion, all these code samples would be functionally equivalent, where a fake getter/setter are created for the property, where you can make assertions about them. When restored, the property descriptor would be changed back to what it was originally.

Reply:
On this point, I think another improvement that would be cool is to support spying on a property. Where the difference would basically be that at the end, if the original property descriptor had a value instead of a getter/setter, we would be changing the property descriptor to having the value be whatever we last set it to while it was being spied on. While a stub just always returns it back to it's original value.